File: 00_Hello_World/Haskell/app/Main.hs
module Main where

import System.Environment (getArgs)
import System.IO (hFlush, stdout)

main :: IO ()
main = do
  args <- getArgs
  case args of
    []           -> sayHello "World"
    ["-i"]       -> prompt "Input your name: " >>= sayHello
    [_]          -> printError $ Just "Unexpected argument on first place, -i expected"
    ["-n", name] -> sayHello name
    [_, _]       -> printError $ Just "Unexpected argument on first place, -n expected"
    _            -> printError $ Just "Invalid arguments."

  where
    sayHello :: String -> IO ()
    sayHello name = putStrLn $ "Hello, " ++ name ++ "!"
    
    prompt :: String -> IO String
    prompt text = do
        putStr text
        hFlush stdout
        getLine
    
    printError :: Maybe String -> IO ()
    printError (Just msg) = putStrLn msg >> printError Nothing
    printError Nothing    = putStr $
      "Use:                                                  \n" ++
      "                                                      \n" ++
      "  hello_world [...args]                               \n" ++
      "                                                      \n" ++
      "  without arguments, print \"Hello, World!\"          \n" ++
      "                                                      \n" ++
      "  -n <name> - print \"Hello, <name>!\"                \n" ++
      "  -i        - ask <name> and print \"Hello, <name>!\" \n"

File: 00_Hello_World/Rust/src/main.rs
use indoc::indoc;
use std::env;
use std::io::Write;

fn say_hello(name: &str) {
    println!("Hello, {}!", name);
}

fn print_error(msg: Option<&str>) {
    if let Some(msg) = msg {
        println!("{}", msg);
    }
    print!(
        indoc! {r#"
            Use:
               hello_world [...args]

               without arguments, print "Hello, World!"

               -n <name> - print "Hello, <name>!"
               -i        - ask <name> and print "Hello, <name>!"
        "#}
    )
}

fn read_line(prefix: Option<&str>) -> String {
    if let Some(prefix) = prefix {
        print!("{}", prefix);
        std::io::stdout().flush().unwrap();
    }
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    input.trim().to_string()
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.is_empty() {
        panic!("Undefined OS, expect that first argument is my path")
    }
    match args[1..] {
        [] =>
            {
                say_hello("World");
            }
        [ref key] =>
            {
                if *key != "-i" {
                    print_error(Some("Unexpected argument on first place, -i expected."));
                    return;
                }
                let name = read_line(Some("Input your name: "));
                say_hello(name.as_str());
            }
        [ref key, ref name] =>
            {
                if *key != "-n" {
                    print_error(Some("Unexpected argument on first place, -n expected."));
                    return;
                }
                say_hello(name.as_str());
            }
        _ =>
            {
                print_error(Some("Unexpected count of arguments."));
                return;
            }
    }
}
